Elasticsearch indices data source

jdesnoes opened the pull request:
Pull request to add an indices data source

tobio commented:
@jdesnoes are you able to take a look at the CLA so we can move forward with this?

jdesnoes commented:
> @jdesnoes are you able to take a look at the CLA so we can move forward with this?

Hello @tobio, I already signed the CLA the 18th of July, with my GitHub's username and email.
But it still looks unchecked on GitHub. And I just signed it again now but same.
Can you take a look? Thanks.

tobio commented:
> But it still looks unchecked on GitHub

I can see you've signed the CLA, however the email address used to author the commits here isn't associated with your GH account at all. If you can re-author the commits with the email shared as part of the CLA process (***@so****ria.com) or any email that's linked to your GH account then it'll get matched up properly.

jdesnoes commented:
> > But it still looks unchecked on GitHub
> 
> I can see you've signed the CLA, however the email address used to author the commits here isn't associated with your GH account at all. If you can re-author the commits with the email shared as part of the CLA process (_**@so**_*ria.com) or any email that's linked to your GH account then it'll get matched up properly.

Thank you, that was the problem indeed, I fixed it and the CLA now appears signed.

jdesnoes commented:
Hello @tobio
I updated my PR with the last changes of the main branch.
Can you take a look and let me know if it's valid to be merged?
Thank you.

tobio commented on internal/elasticsearch/index/indices/read.go:
```diff
@@ -0,0 +1,179 @@
+			resp.Diagnostics.AddError("unable to populate index from settings map", err.Error())
+			return
+		}
+		// TODO: We ideally should set read settings to each field to detect changes
+		// But for now, setting it will cause unexpected diff for the existing clients which use `settings`
```
Is this comment still accurate, IIUC `populateIndexFromSettings` will do exactly what this comment is talking about?

tobio commented:
@jdesnoes sorry, I've been meaning to look at this in comparison to https://github.com/elastic/terraform-provider-elasticstack/pull/698 which moves the existing index resource to the plugin framework being used here. We should be able to re-use a lot of the model code in both of these PR's. 

I think it makes sense to get this merged and then I can merge the two code bases together. Just a heads up, I think it makes sense to get both of these merged quite close together. This one may sit approved but unmerged for a little while whilst I get them both inline. 

The actual code changes look good, thanks for adding this! Are you able to take a look at why the tests are failing?

jdesnoes commented:
> @jdesnoes sorry, I've been meaning to look at this in comparison to #698 which moves the existing index resource to the plugin framework being used here. We should be able to re-use a lot of the model code in both of these PR's.
> 
> I think it makes sense to get this merged and then I can merge the two code bases together. Just a heads up, I think it makes sense to get both of these merged quite close together. This one may sit approved but unmerged for a little while whilst I get them both inline.
> 
> The actual code changes look good, thanks for adding this! Are you able to take a look at why the tests are failing?

Thanks for your answer @tobio 

Indeed, considering the #698 PR purpose, there is some code that can be used in common with this current PR, mostly for the model.
Considering that, and to make the merge between both of those pull requests easier, I already updated a bit my code to make it more similar to yours, mainly the models and the schema.

I also fixed the unit tests.

So that, no problem if in the meanwhile this PR has to be sit approved but unmerged.

tobio commented:
Thanks mate, we're working through the review process on the other one now, so we should get the both merged in the next couple of weeks.

tobio commented on internal/elasticsearch/index/indices/schema.go:
```diff
@@ -0,0 +1,434 @@
+				Description: "Generated ID for the indices.",
+				Computed:    true,
+			},
+			"search": schema.StringAttribute{
+				Description: "Comma-separated list of indices to resolve by their name. Supports wildcards `*`.",
+				Optional:    true,
+			},
```
```suggestion
			"target": schema.StringAttribute{
				Description: "Comma-separated list of data streams, indices, and aliases used to limit the request. Supports wildcards (*). To target all data streams and indices, omit this attribute or use * or _all. ",
				Optional:    true,
			},
```

Copying the details from the ES [docs](https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-get-index.html#get-index-api-path-params). I'm not a huge fan of the `search` name, and would have thought of this as `name` or `name_pattern` or similar but in reality it could also be an alias resolving to multiple indices so sticking with `target` might make the most sense. WDYT?

tobio reviewed (commented):
Thanks for your patience on this one. One small change on the schema, where I think keeping with the documented names may be nicer. Otherwise this LGTM.

jdesnoes commented:
Thanks again for your help and your review @tobio 

You're definitely right, it makes sense to keep the name and the description of the official Elasticsearch API documentation.
Indeed, `target` looks the best option to cover all the cases as the search field name.

I just updated the pull request to take this in account, thanks.

tobio commented:
I'll look at merging the duplicate code in a follow up PR. Thanks for taking the time to add this one in @jdesnoes!
